File: buerguer-pig/HamburgesasEIngredientes-Java-IvanRC/src/main/java/storage/ingrediente/IngredienteStorageServiceBinario.java
package storage.ingrediente;

import config.ConfigApp;
import models.Ingrediente;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class IngredienteStorageServiceBinario implements IngredienteStorageService{

    ConfigApp configApp = ConfigApp.getInstance();

    File file = new File(configApp.getAPP_DATA()+File.separator+"ingrediente.bin");

    public IngredienteStorageServiceBinario() throws IOException {
    }

    @Override
    public void saveAll(List<Ingrediente> entities) throws Exception {
        try(FileOutputStream fileOutputStream = new FileOutputStream(file)){
            entities.forEach(ingrediente -> {
                    try {
                        fileOutputStream.write((ingrediente.getId()+"\n").getBytes());
                        fileOutputStream.write((ingrediente.getNombre()+"\n").getBytes());
                        fileOutputStream.write((ingrediente.getPrecio()+"\n").getBytes());
                    }catch (Exception e) {
                        System.out.println(e.getMessage());
                    }
            }
            );
        }
    }

    @Override
    public List<Ingrediente> loadAll() throws Exception {
        List<Ingrediente> ingredientes = new ArrayList<>();
        if(!file.exists()) return ingredientes;
        try(FileInputStream fileInputStream = new FileInputStream(file)){
            char c = ' ';

            do {
                StringBuilder id = new StringBuilder();
                if(c != ' ') {
                    do {
                        id.append(c);
                    }while((c = (char) fileInputStream.read()) != '\n');
                }else{
                    do{
                        c = (char) fileInputStream.read();
                        id.append(c);
                    }while ((c = (char) fileInputStream.read()) != '\n');
                }

                StringBuilder nombre = new StringBuilder();
                while ((c = (char) fileInputStream.read()) != '\n') {
                    nombre.append((char) c);
                }

                StringBuilder precio = new StringBuilder();
                while ((c = (char) fileInputStream.read()) != '\n') {
                    precio.append((char) c);
                }

                ingredientes.add(
                        new Ingrediente(
                                Integer.parseInt(String.valueOf(id)),
                                nombre.toString(),
                                Double.parseDouble(precio.toString())
                        )
                );
                c = (char) fileInputStream.read();
            }while(c != '\uFFFF');
        }
        return ingredientes;
    }
}

File: buerguer-pig/HamburgesasEIngredientes-Kotlin-IvanRC/src/main/kotlin/storageService/Ingrediente/IngredienteStorageServiceBinario.kt
package storageService.Ingrediente

import config.ConfigApp
import model.Ingrediente
import mu.KotlinLogging
import java.io.File

class IngredienteStorageServiceBinario : IngredienteStorageService {

    private val config = ConfigApp
    private val file = File(config.APP_DATA+ File.separator+"ingredientes.bin")
    private val logger = KotlinLogging.logger {}

    override fun saveAll(entities: List<Ingrediente>) {
        logger.debug { "Se guardan todos los ingredientes en el fichero binario." }
        file.writeBytes("".toByteArray())
        entities.forEach { ingrediente ->
            file.appendBytes((ingrediente.id.toString()+"\n").toByteArray())
            file.appendBytes((ingrediente.nombre+"\n").toByteArray())
            file.appendBytes((ingrediente.precio.toString()+"\n").toByteArray())
        }
    }

    override fun loadAll(): List<Ingrediente> {
        logger.debug { "Se cargán todos los ingredientes del fichero binario." }
        var ingredientes = mutableListOf<Ingrediente>()
        if(!file.exists()) return ingredientes.toList()
        file.inputStream().use {
            var char = it.read().toChar()

            val id = StringBuilder()
            while(char != '\n'){
                id.append(char)
                char = it.read().toChar()
            }

            char = it.read().toChar()

            val nombre = StringBuilder()
            while(char != '\n'){
                nombre.append(char)
                char = it.read().toChar()
            }

            char = it.read().toChar()

            val precio = StringBuilder()
            while(char != '\n'){
                precio.append(char)
                char = it.read().toChar()
            }

            ingredientes.add(
                Ingrediente(
                    id = id.toString().toInt(),
                    nombre = nombre.toString(),
                    precio = precio.toString().toDouble()
                )
            )
        }
        return ingredientes.toList()
    }

}